File: src/main/java/be/angelcorp/libs/celest/potential/CumulativePotential.java
/**
 * Copyright (C) 2009-2012 simon <simon@angelcorp.be>
 *
 * Licensed under the Non-Profit Open Software License version 3.0
 * (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *        http://www.opensource.org/licenses/NOSL3.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package be.angelcorp.libs.celest.potential;

import java.util.Arrays;
import java.util.List;

import be.angelcorp.libs.math.linear.Vector3D;

import com.google.common.collect.Lists;

/**
 * Implementation of {@link ICumulativePotential}
 * 
 * @author simon
 * @see ICumulativePotential
 */
public class CumulativePotential implements ICumulativePotential {

	/**
	 * List containing potentials that are to be summed
	 */
	private List<IGravitationalPotential>	potentials;

	/**
	 * Create a potential resulting from a set of other gravity potentials
	 * 
	 * @param potentials
	 *            List of other potentials
	 */
	public CumulativePotential(IGravitationalPotential... potentials) {
		this.potentials = Lists.newLinkedList(Arrays.asList(potentials));

	}

	/**
	 * Create a potential resulting from a set of other gravity potentials
	 * 
	 * @param potentials
	 *            List of other potentials
	 */
	public CumulativePotential(List<IGravitationalPotential> potentials) {
		this.potentials = potentials;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Vector3D evaluate(Vector3D point) {
		Vector3D sum = new Vector3D(0, 0, 0);
		for (IGravitationalPotential pot : potentials) {
			sum = sum.add(pot.evaluate(point));
		}
		return sum;
	}

}

File: celest/src/main/java/be/angelcorp/celest/potential/CumulativePotential.scala
/**
 * Copyright (C) 2009-2012 simon <simon@angelcorp.be>
 *
 * Licensed under the Non-Profit Open Software License version 3.0
 * (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.opensource.org/licenses/NOSL3.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package be.angelcorp.celest.potential

import be.angelcorp.celest.math.geometry.Vec3

/**
 * Create a GravitationalPotential that is the result of the summation of a set of independent
 * GravitationalPotentials
 *
 * @param potentials List containing potentials that are to be summed.
 *
 * @author Simon Billemont
 */
class CumulativePotential(val potentials: List[GravitationalPotential]) extends GravitationalPotential {

  override def apply(point: Vec3) =
    potentials.foldLeft(Vec3.zero)((sum, p) => sum + p(point))

}
